### ISSUE TITLE
[Bug] Karpenter fails to provision

### ISSUE THREAD

**Issue by rothgar** [weaveworks/eksctl#6033]
### What were you trying to accomplish?

Created a cluster with the following configuration

```yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: karp
  region: us-east-2
  version: '1.24'
  tags:
    karpenter.sh/discovery: karp

iam:
  withOIDC: true

karpenter:
  version: 'v0.20.0'
  createServiceAccount: true

managedNodeGroups:
  - name: addons-ng-1
    minSize: 2
    maxSize: 5
    desiredCapacity: 2
    taints:
      - key: CriticalAddonsOnly
        value: "true"
        effect: NoSchedule

addons:
  - name: coredns
  - name: aws-ebs-csi-driver
    wellKnownPolicies:
      ebsCSIController: true
  - name: vpc-cni
```

### What happened?

The cluster created with karpenter installed but pods were crashing because they were missing settings. The karpenter-global-settings configmap in the karpenter namespace was missing the settings for 

```
   aws.clusterEndpoint: ""
   aws.clusterName: ""
   aws.defaultInstanceProfile: ""
```
After filling in this information the controller would run but was giving the following error in the logs
```
ERROR   webhook.ValidationWebhook       Reconcile error {"commit": "f60dacd", "knative.dev/traceid": "5680f5d4-871
b-46e2-a32e-9808d5b88fdb", "knative.dev/key": "karpenter/karpenter-cert", "duration": "115.870254ms", "error": "failed to update webhook: Operatio
n cannot be fulfilled on validatingwebhookconfigurations.admissionregistration.k8s.io \"validation.webhook.karpenter.k8s.aws\": the object has bee
n modified; please apply your changes to the latest version and try again"}
```
I had to apply the latest karpenter CRDs to the cluster with
```
kubectl replace -f https://raw.githubusercontent.com/aws/karpenter/v0.20.0/pkg/apis/crds/karpenter.sh_provisioners.yaml
kubectl replace -f https://raw.githubusercontent.com/aws/karpenter/v0.20.0/pkg/apis/crds/karpenter.k8s.aws_awsnodetemplates.yaml
```
At this point I could create the karpenter provisioner and AWSNodeTemplate and then I was ready to 

### How to reproduce it?

Use the config provided with `eksctl create cluster -f config.yaml`

**Logs**

**Anything else we need to know?**

**Versions**

```
$ eksctl info
eksctl version: 0.122.0-dev+b7686d461.2022-11-29T18:34:06Z
kubectl version: v1.25.4
OS: linux
```

**Comment by ryann-sportsbet** [comment]
I have the same issue. Did you try the lower version?

**Comment by rothgar** [comment]
Do you mean a lower version of karpenter, eksctl, or eks?

**Comment by marcossv9** [comment]
I've got the same issue while using eks `v1.24` and karpenter `v0.20.0`. Here my eksctl and kubectl versions:
```
eksctl version: 0.124.0
kubectl version: v1.24.7-eks-fb459a0
OS: linux
```

Here are the logs from the carpenter controller:

```
kubectl logs -n karpenter karpenter-645754498f-jx7sp controller --follow
2022-12-28T17:50:18.417Z        DEBUG   Successfully created the logger.
2022-12-28T17:50:18.417Z        DEBUG   Logging level set to: debug
{"level":"info","ts":1672249818.421702,"logger":"fallback","caller":"injection/injection.go:63","msg":"Starting informers..."}
2022-12-28T17:50:18.522Z        DEBUG   controller      waiting for configmaps  {"commit": "f60dacd", "configmaps": ["karpenter-global-settings"]}
E1228 17:50:19.026538       1 runtime.go:77] Observed a panic: validating settings, "" not a valid clusterEndpoint URL; Key: 'Settings.ClusterName' Error:Field validation for 'ClusterName' failed on the 'required' tag
Key: 'Settings.ClusterEndpoint' Error:Field validation for 'ClusterEndpoint' failed on the 'required' tag
goroutine 126 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x213fae0?, 0xc000367c90})
        k8s.io/apimachinery@v0.25.4/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0009831f0?})
        k8s.io/apimachinery@v0.25.4/pkg/util/runtime/runtime.go:49 +0x75
panic({0x213fae0, 0xc000367c90})
        runtime/panic.go:884 +0x212
github.com/aws/karpenter/pkg/apis/config/settings.NewSettingsFromConfigMap(0xc000825200)
        github.com/aws/karpenter/pkg/apis/config/settings/settings.go:93 +0x4fb
reflect.Value.call({0x228dac0?, 0x2865dc8?, 0x0?}, {0x274531b, 0x4}, {0xc000989c40, 0x1, 0x0?})
        reflect/value.go:584 +0x8c5
reflect.Value.Call({0x228dac0?, 0x2865dc8?, 0xc000588fa0?}, {0xc000989c40?, 0x203000?, 0x203000?})
        reflect/value.go:368 +0xbc
knative.dev/pkg/configmap.(*UntypedStore).OnConfigChanged(0xc0003fdcc0, 0xc000825200)
        knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/store.go:141 +0x185
knative.dev/pkg/configmap.(*ManualWatcher).OnChange(0xc0004d19b0, 0xc000825200)
        knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/manual_watcher.go:72 +0x106
knative.dev/pkg/configmap/informer.(*InformedWatcher).addConfigMapEvent(0x740000000008?, {0x270b2c0?, 0xc000825200?})
        knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/informer/informed_watcher.go:220 +0x38
knative.dev/pkg/configmap/informer.(*syncedCallback).Call(0xc00087a700, {0x270b2c0?, 0xc000825200?}, {0xc000588fa0, 0x19})
        knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/informer/synced_callback.go:94 +0x3d
knative.dev/pkg/configmap/informer.(*InformedWatcher).Start.func1({0x270b2c0?, 0xc000825200?})
        knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/informer/informed_watcher.go:158 +0x45
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
        k8s.io/client-go@v0.25.4/tools/cache/controller.go:232
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        k8s.io/client-go@v0.25.4/tools/cache/shared_informer.go:818 +0x134
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x406d01?)
        k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000071f38?, {0x2c55f40, 0xc0008ecf00}, 0x1, 0xc000b361e0)
        k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x7b7ed0?, 0x3b9aca00, 0x0, 0x26?, 0xc000071f88?)
        k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:92
k8s.io/client-go/tools/cache.(*processorListener).run(0xc00017e800?)
        k8s.io/client-go@v0.25.4/tools/cache/shared_informer.go:812 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
        k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:75 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
        k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:73 +0x85
panic: validating settings, "" not a valid clusterEndpoint URL; Key: 'Settings.ClusterName' Error:Field validation for 'ClusterName' failed on the 'required' tag
Key: 'Settings.ClusterEndpoint' Error:Field validation for 'ClusterEndpoint' failed on the 'required' tag [recovered]
        panic: validating settings, "" not a valid clusterEndpoint URL; Key: 'Settings.ClusterName' Error:Field validation for 'ClusterName' failed on the 'required' tag
Key: 'Settings.ClusterEndpoint' Error:Field validation for 'ClusterEndpoint' failed on the 'required' tag

goroutine 126 [running]:
```
I'll test it using a lower version of karpenter and post the results.

**Comment by marcossv9** [comment]
Tested same settings with lower version of karpenter, `v0.19.3`. No luck.

**Comment by marcossv9** [comment]
It worked with karpenter `v0.18.1`, something has to be broken since `v0.19.3`.

**Comment by emboss64** [comment]
Karpenter changed the key names to settings.aws.XXX as of v0.19.x:
- clusterName > settings.aws.clusterName
- clusterEndpoint > settings.aws.clusterEndpoint
- aws.defaultInstanceProfile > settings.aws.defaultInstanceProfile
- X > settings.aws.interruptionQueueName (*this one is new, so also worth checking if eksctl will create the needed SQS queue needed by karpenter*)

**Comment by Himangini** [comment]
From `v0.19.0` onwards few key things have changed in Karpenter releases as also pointed out by @emboss64. I'll aim to add a fix for the latest versions around next week, till then I'd suggest sticking to `v0.18.1` or lower for the time being. 😓

**Comment by colstrom** [comment]
For what it's worth, I was able to get Karpenter `v0.22.1` to finish provisioning by patching the `ConfigMap/karpenter-global-settings` in the `karpenter` namespace while `eksctl` was waiting at the `adding Karpenter to cluster ${CLUSTER_NAME}` stage, using the following:

Find `CLUSTER_ENDPOINT` with:
```shell
aws eks describe-cluster --name ${CLUSTER_NAME} | jq -r .cluster.endpoint
```

Replace `CLUSTER_NAME` and `CLUSTER_ENDPOINT` in the following YAML:
```yaml
# patch.yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: karpenter-global-settings
  namespace: karpenter
data:
  aws.clusterEndpoint: ${CLUSTER_ENDPOINT}
  aws.clusterName: ${CLUSTER_NAME}
  aws.defaultInstanceProfile: eksctl-KarpenterNodeInstanceProfile-${CLUSTER_NAME}
```

Apply the patch:
```shell
kubectl patch \
  --patch-file=patch.yaml \
  --namespace=karpenter \
  ConfigMap/karpenter-global-settings
```

I haven't tested it beyond that yet, but it completes provisioning and ends up in a `Ready` state, at least.

**Comment by ruzickap** [comment]
Hello guys.
I tried the `eksctl` version `0.126.0` with karpenter configuration:

```yaml
karpenter:
  version: v0.20.0
  # same problem is with v0.22.1
```

... and I'm still getting:

```
❯ kubectl get pods -A
NAMESPACE     NAME                                READY   STATUS             RESTARTS       AGE
karpenter     karpenter-645754498f-gcmtq          0/1     CrashLoopBackOff   56 (27s ago)   4h22m
karpenter     karpenter-645754498f-jv6dw          0/1     CrashLoopBackOff   56 (64s ago)   4h22m
```

```bash
❯ kubectl logs -n karpenter karpenter-645754498f-gcmtq
2023-01-21T19:23:05.571Z	DEBUG	Successfully created the logger.
2023-01-21T19:23:05.571Z	DEBUG	Logging level set to: debug
{"level":"info","ts":1674328985.57504,"logger":"fallback","caller":"injection/injection.go:63","msg":"Starting informers..."}
2023-01-21T19:23:05.676Z	DEBUG	controller	waiting for configmaps	{"commit": "f60dacd", "configmaps": ["karpenter-global-settings"]}
2023-01-21T19:23:06.180Z	DEBUG	controller	karpenter-global-settings config "karpenter-global-settings" config was added or updated: settings.Settings{BatchMaxDuration:v1.Duration{Duration:10000000000}, BatchIdleDuration:v1.Duration{Duration:1000000000}}	{"commit": "f60dacd"}
E0121 19:23:06.181000       1 runtime.go:77] Observed a panic: validating settings, "" not a valid clusterEndpoint URL; Key: 'Settings.ClusterName' Error:Field validation for 'ClusterName' failed on the 'required' tag
Key: 'Settings.ClusterEndpoint' Error:Field validation for 'ClusterEndpoint' failed on the 'required' tag
goroutine 133 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1b9ca00?, 0x40005675c0})
	k8s.io/apimachinery@v0.25.4/pkg/util/runtime/runtime.go:75 +0x84
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x40007390a8?})
	k8s.io/apimachinery@v0.25.4/pkg/util/runtime/runtime.go:49 +0x80
panic({0x1b9ca00, 0x40005675c0})
	runtime/panic.go:884 +0x20c
github.com/aws/karpenter/pkg/apis/config/settings.NewSettingsFromConfigMap(0x400093cea0)
	github.com/aws/karpenter/pkg/apis/config/settings/settings.go:93 +0x408
reflect.Value.call({0x1cea8c0?, 0x22c29c0?, 0x4000739af8?}, {0x21a2032, 0x4}, {0x4000739c00, 0x1, 0x1ed7180?})
	reflect/value.go:584 +0x688
reflect.Value.Call({0x1cea8c0?, 0x22c29c0?, 0x400077d380?}, {0x4000739c00?, 0x400078dbf8?, 0x2f5068?})
	reflect/value.go:368 +0x90
knative.dev/pkg/configmap.(*UntypedStore).OnConfigChanged(0x40004c2d20, 0x400093cea0)
	knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/store.go:141 +0x15c
knative.dev/pkg/configmap.(*ManualWatcher).OnChange(0x4000954a50, 0x400093cea0)
	knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/manual_watcher.go:72 +0x120
knative.dev/pkg/configmap/informer.(*InformedWatcher).addConfigMapEvent(0x40001683c0?, {0x2167fe0?, 0x400093cea0?})
	knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/informer/informed_watcher.go:220 +0x40
knative.dev/pkg/configmap/informer.(*syncedCallback).Call(0x4000137de0, {0x2167fe0?, 0x400093cea0?}, {0x400077d380, 0x19})
	knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/informer/synced_callback.go:94 +0x40
knative.dev/pkg/configmap/informer.(*InformedWatcher).Start.func1({0x2167fe0?, 0x400093cea0?})
	knative.dev/pkg@v0.0.0-20221123154742-05b694ec4d3a/configmap/informer/informed_watcher.go:158 +0x4c
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
	k8s.io/client-go@v0.25.4/tools/cache/controller.go:232
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	k8s.io/client-go@v0.25.4/tools/cache/shared_informer.go:818 +0x13c
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x9200?)
	k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:157 +0x40
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x400006ef28?, {0x26b2c60, 0x40006e8510}, 0x1, 0x4000167f80)
	k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:158 +0x90
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x400006ef58?, 0x3b9aca00, 0x0, 0x88?, 0x400006ef78?)
	k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:135 +0x80
k8s.io/apimachinery/pkg/util/wait.Until(...)
	k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:92
k8s.io/client-go/tools/cache.(*processorListener).run(0x4000942a00?)
	k8s.io/client-go@v0.25.4/tools/cache/shared_informer.go:812 +0x68
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:75 +0x58
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
 1 delx                                                                                                                                                                                                                                                                                            Buffers
	k8s.io/apimachinery@v0.25.4/pkg/util/wait/wait.go:73 +0x84
panic: validating settings, "" not a valid clusterEndpoint URL; Key: 'Settings.ClusterName' Error:Field validation for 'ClusterName' failed on the 'required' tag
Key: 'Settings.ClusterEndpoint' Error:Field validation for 'ClusterEndpoint' failed on the 'required' tag [recovered]
	panic: validating settings, "" not a valid clusterEndpoint URL; Key: 'Settings.ClusterName' Error:Field validation for 'ClusterName' failed on the 'required' tag
Key: 'Settings.ClusterEndpoint' Error:Field validation for 'ClusterEndpoint' failed on the 'required' tag
```

The patch provided by @colstrom is working... 
The bug should be reopened, because it is still valid...

**Comment by emboss64** [comment]
That's because the bugfix is not included in 0.126.0 :)
You can test it in [v0.127.0-rc.0](https://github.com/weaveworks/eksctl/releases/tag/v0.127.0-rc.0)

**Comment by ruzickap** [comment]
You are right...
My bad ... sorry...

In `eksctl` version [v0.127.0-rc.0](https://github.com/weaveworks/eksctl/releases/tag/v0.127.0-rc.0) karpenter is working in latest version (`0.22.1`).
